util/check_tool_requirements.py

```diff
@@ -23,6 +25,277 @@ def get_tool_requirements_path():
+                    'Bad semver inferred from required version ({}): {}'
+                    .format(min_semver, err))
+
+        try:
+            actual_version = self.get_version()
+        except RuntimeError as err:
+            return (False, str(err))
+
+        try:
+            actual_semver = self.to_semver(actual_version)
+        except ValueError as err:
+            return (False,
+                    'Failed to convert installed to semantic version: {}'
+                    .format(err))
+        try:
+            actual_sv = StrictVersion(actual_semver)
+        except ValueError as err:
+            return (False,
+                    'Bad semver inferred from installed version ({}): {}'
+                    .format(actual_semver, err))
+
+        if actual_sv < min_sv:
+            return (False,
+                    'Installed version is too old: '
+                    'found version {}, but need at least {}'
+                    .format(actual_version, self.min_version))
+
+        return (True,
+                'Sufficiently recent version (found {}; needed {})'
+                .format(actual_version, self.min_version))
+
+
+class VerilatorToolReq(ToolReq):
+    def get_version(self):
+        try:
+            # Note: "verilator" needs to be called through a shell and with all
+            # arguments in a string, as it doesn't have a shebang, but instead
+            # relies on perl magic to parse command line arguments.
+            version_str = subprocess.run('verilator --version', shell=True,
+                                         check=True, stdout=subprocess.PIPE,
+                                         stderr=subprocess.STDOUT,
+                                         universal_newlines=True)
+        except subprocess.CalledProcessError as err:
+            raise RuntimeError('Unable to call Verilator to check version: {}'
+                               .format(err)) from None
+
+        return version_str.stdout.split(' ')[1].strip()
+
+
+class VeribleToolReq(ToolReq):
+    tool_cmd = ['verible-verilog-lint', '--version']
+
+    def to_semver(self, version):
+        # Drop the hash suffix and convert into version string that
+        # is compatible with StrictVersion in check_version below.
+        # Example: v0.0-808-g1e17daa -> 0.0.808
+        m = re.fullmatch(r'v([0-9]+)\.([0-9]+)-([0-9]+)-g[0-9a-f]+$', version)
+        if m is None:
+            raise ValueError("{} has invalid version string format."
+                             .format(version))
+
+        return '.'.join(m.group(1, 2, 3))
+
+
+class VcsToolReq(ToolReq):
+    tool_cmd = ['vcs', '-full64', '-ID']
+    tool_env = {'VCS_ARCH_OVERRIDE': 'linux'}
+    version_regex = re.compile(r'Compiler version = VCS [A-Z]-(.*)')
+
+    def to_semver(self, version):
+        # VCS versions look like 2020.03-SP1-blahblahblah. Keep the year,
+        # month and SP numbers.
+        match = re.match(r'([0-9]+).([0-9]+)(?:-SP([0-9]+))?', version)
```

Comment:
I indicated in a previous comment that there are incremental SP patches as well. Example: `2020.03-SP2-2`.

Reply:
Hmm, yes, you're right. And in fact, there are also things like Q-2020.03-1 (which came after Q-2020.03-SP1, but before Q-2020.03-SP1-1: clearly, lexicographical ordering is for the weak!)

I'll translate like this: Q-2020.03-SP1-2 => 2020.03.201 (where the "-2" turns into 2 * 100, to cram the versioning into StrictVersion's 3 digit scheme).